lib-ee/emqx_license/src/emqx_license_parser_v20220101.erl

```diff
@@ -0,0 +1,137 @@
+    Expiry = DateNow > DateExpiry,
+
+    [{customer, Customer},
+     {email, Email},
+     {max_connections, MaxConns},
+     {start_at, format_date(DateStart)},
+     {expiry_at, format_date(DateExpiry)},
+     {type, format_type(Type)},
+     {customer_type, CType},
+     {expiry, Expiry}].
+
+customer_type(#{customer_type := CType}) -> CType.
+
+license_type(#{type := Type}) -> Type.
+
+expiry_date(#{date_start := DateStart, days := Days}) ->
+    calendar:gregorian_days_to_date(
+     calendar:date_to_gregorian_days(DateStart) + Days).
+
+max_connections(#{max_connections := MaxConns}) ->
+    MaxConns.
+
+%%------------------------------------------------------------------------------
+%% Private functions
+%%------------------------------------------------------------------------------
+
+verify_signature(Payload, Signature, Key) ->
+    RSAPublicKey = public_key:der_decode('RSAPublicKey', Key),
+    public_key:verify(Payload, ?DIGEST_TYPE, Signature, RSAPublicKey).
+
+parse_payload(Payload) ->
+    Lines = lists:map(
+              fun string:trim/1,
+              string:split(string:trim(Payload), <<"\n">>, all)),
+    case Lines of
+        [?LICENSE_VERSION, Type, CType, Customer, Email, DateStart, Days, MaxConns] ->
+            collect_fields([{type, parse_type(Type)},
+                            {customer_type, parse_customer_type(CType)},
+                            {customer, {ok, Customer}},
+                            {email, {ok, Email}},
+                            {date_start, parse_date_start(DateStart)},
+                            {days, parse_days(Days)},
+                            {max_connections, parse_max_connections(MaxConns)}]);
+        _ ->
+            {error, invalid_format}
+    end.
+
+parse_type(TypeStr) ->
+    case string:to_integer(TypeStr) of
+        {Type, <<"">>} -> {ok, Type};
+        _ -> {error, invalid_license_type}
+    end.
+
+parse_customer_type(CTypeStr) ->
+    case string:to_integer(CTypeStr) of
+        {CType, <<"">>} -> {ok, CType};
+        _ -> {error, invalid_customer_type}
+    end.
+
+parse_date_start(<<Y:4/binary, M:2/binary, D:2/binary>>) ->
+    Date = list_to_tuple([N || {N, <<>>} <- [string:to_integer(S) || S <- [Y, M, D]]]),
+    case calendar:valid_date(Date) of
+        true -> {ok, Date};
+        false -> {error, invalid_date_start}
+    end;
+parse_date_start(_) -> {error, invalid_date_start}.
+
+parse_days(DaysStr) ->
+    case string:to_integer(DaysStr) of
+        {Days, <<"">>} when Days > 0 -> {ok, Days};
+        _ -> {error, invalid_day_period}
+    end.
+
+parse_max_connections(MaxConnStr) ->
+    case string:to_integer(MaxConnStr) of
+        {MaxConns, <<"">>} when MaxConns > 0 -> {ok, MaxConns};
+        _ -> {error, invalid_max_connections}
+    end.
+
+collect_fields(Fields) ->
+    lists:foldl(
+      fun({Name, {ok, Value}}, {ok, Map}) -> {ok, Map#{Name => Value}};
+         (_, {error, _} = Error) -> Error;
```

Comment:
Nit: maybe we can collect and display all errors instead of stopping at the first one, since the validation work is already done.

Reply:
✅